### PR TITLE
Bugfixes and performance improvements for parallel_hash

### PR DESCRIPTION
**Pull request by vdimir**
### Changelog category (leave one):

- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

* Bugfixes and performance improvements for `parallel_hash`

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by lgbo-ustc** [comment]
@amosbird Do you have the sql to create table `date` ? I don't find it in https://clickhouse.com/docs/zh/getting-started/example-datasets/star-schema/

**Comment by vdimir** [comment]
Slow down may happen if we select lots of columns from left table, because it'll dispatch (call `scatter`) all of them. Not solved yet. 
So, not sure about enabling, but this PR fixes some bugs found by running parallel_hash on all tests set.

**Comment by BigRedEye** [comment]
I think I discovered a rather subtle bug while working on the related code at #38191. The bug was introduced in d7129855757f38ceec3e4ecc6dafacdabe9b178f with transition from SipHash to WeakHash32 for sharding: the hash tables within buckets use the last bits of CRC32 as hash value as well. Thus, the distribution of hash values within each bucket is skewed, the number of collisions increases significantly, and the performance of hash tables degrades.
Maybe we should use an orthogonal hash function for sharding?

**Comment by vdimir** [comment]
I decided not to set it as a default for now, but running this algorithm on a complete test set helped to find some bugs and fix them.

File: src/Interpreters/ConcurrentHashJoin.cpp [hunk 1]
```diff
@@ -26,18 +27,23 @@ namespace ErrorCodes
 {
     extern const int LOGICAL_ERROR;
     extern const int SET_SIZE_LIMIT_EXCEEDED;
-    extern const int BAD_ARGUMENTS;
+}
+
+static UInt32 toPowerOfTwo(UInt32 x)
+{
+    x = x - 1;
```
Comment (nickitat) [inline src/Interpreters/ConcurrentHashJoin.cpp:34]: minor, but maybe just `static_cast<UInt32>(1) << (31 - __builtin_clz(x))`

File: src/Interpreters/ConcurrentHashJoin.cpp [hunk 2]
```diff
@@ -26,18 +27,23 @@ namespace ErrorCodes
...
     : context(context_)
     , table_join(table_join_)
     , slots(slots_)
 {
-    if (slots < 1 || 255 < slots)
-    {
-        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Number of slots should be [1, 255], got {}", slots);
-    }
+    slots = std::min<size_t>(std::max<size_t>(1, slots), 255);
+    slots = toPowerOfTwo(slots);
```
Comment (nickitat) [inline src/Interpreters/ConcurrentHashJoin.cpp:46]: maybe place these two inside ctor in l.43

**Review by nickitat — APPROVED** [review]
lgtm, just minor remarks

**Comment by vdimir** [comment]
> Stress test (address, actions) — Download script failed
> Stress test (debug, actions) — Download script failed
> Stress test (memory, actions) — Download script failed
> Stress test (thread, actions) — Download script failed
> Stress test (undefined, actions) — Download script failed

Looks suspicious

**Pull request merged**